File: tests/src/test/java/org/veri/be/integration/usecase/SocialReadingIntegrationTest.java
package org.veri.be.integration.usecase;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.veri.be.domain.book.entity.Book;
import org.veri.be.domain.book.entity.Reading;
import org.veri.be.domain.book.repository.BookRepository;
import org.veri.be.domain.book.repository.ReadingRepository;
import org.veri.be.integration.IntegrationTestSupport;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

class SocialReadingIntegrationTest extends IntegrationTestSupport {

    @Autowired BookRepository bookRepository;
    @Autowired ReadingRepository readingRepository;

    @Nested
    @DisplayName("GET /api/v2/bookshelf/popular")
    class GetPopular {
        @Test
        @DisplayName("인기 도서 10개 조회")
        void getPopularSuccess() throws Exception {
            mockMvc.perform(get("/api/v2/bookshelf/popular"))
                    .andExpect(status().isOk());
        }

        @Test
        @DisplayName("최근 일주일 내 추가 도서 없음")
        void getPopularEmpty() throws Exception {
            // Assuming clean DB or no recent adds
            mockMvc.perform(get("/api/v2/bookshelf/popular"))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.result.books").isEmpty());
        }

        @Test
        @DisplayName("인기 도서 최대 10개 제한")
        void getPopularLimit() throws Exception {
            // Create 12 readings for 12 different books
            for (int i = 0; i < 12; i++) {
                createReading(true, getMockMember(), "ISBN" + i);
            }

            mockMvc.perform(get("/api/v2/bookshelf/popular"))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.result.books.length()").value(org.hamcrest.Matchers.lessThanOrEqualTo(10)));
        }
    }

    @Nested
    @DisplayName("GET /api/v2/bookshelf/{readingId}")
    class GetReadingDetail {
        @Test
        @DisplayName("공개 독서 상세")
        void getReadingDetailSuccess() throws Exception {
            Reading reading = createReading(true, getMockMember(), "ISBN-DETAIL");

            mockMvc.perform(get("/api/v2/bookshelf/" + reading.getId()))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.result.memberBookId").value(reading.getId()));
        }

        @Test
        @DisplayName("비공개 + 소유자 조회")
        void getPrivateOwner() throws Exception {
            Reading reading = createReading(false, getMockMember(), "ISBN-PRIVATE");

            mockMvc.perform(get("/api/v2/bookshelf/" + reading.getId()))
                    .andExpect(status().isOk());
        }

        @Test
        @DisplayName("비공개 + 타인 접근")
        void getPrivateOther() throws Exception {
            org.veri.be.domain.member.entity.Member other = org.veri.be.domain.member.entity.Member.builder()
                    .email("o").nickname("o").profileImageUrl("p").providerId("p").providerType(org.veri.be.domain.member.entity.enums.ProviderType.KAKAO).build();
            memberRepository.save(other);
            Reading reading = createReading(false, other, "ISBN-OTHER");

            mockMvc.perform(get("/api/v2/bookshelf/" + reading.getId()))
                    .andExpect(status().isNotFound());
        }

        @Test
        @DisplayName("존재하지 않는 ID")
        void getNotFound() throws Exception {
            mockMvc.perform(get("/api/v2/bookshelf/999"))
                    .andExpect(status().isBadRequest()); // As per checklist expectation (BookErrorCode.BAD_REQUEST)
        }
    }

    private Reading createReading(boolean isPublic, org.veri.be.domain.member.entity.Member member, String isbn) {
        Book book = Book.builder().title("T").image("I").isbn(isbn).build();
        book = bookRepository.save(book);
        Reading reading = Reading.builder()
                .member(member)
                .book(book)
                .isPublic(isPublic)
                .build();
        return readingRepository.save(reading);
    }
}

File: tests/src/test/kotlin/org/veri/be/integration/usecase/SocialReadingIntegrationTest.kt
package org.veri.be.integration.usecase

import org.hamcrest.Matchers
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.veri.be.domain.book.repository.BookRepository
import org.veri.be.domain.book.repository.ReadingRepository
import org.veri.be.domain.member.entity.enums.ProviderType
import org.veri.be.integration.IntegrationTestSupport
import org.veri.be.support.fixture.BookFixture
import org.veri.be.support.fixture.MemberFixture
import org.veri.be.support.fixture.ReadingFixture
import org.veri.be.support.steps.SocialReadingSteps

class SocialReadingIntegrationTest : IntegrationTestSupport() {

    @Autowired
    private lateinit var bookRepository: BookRepository

    @Autowired
    private lateinit var readingRepository: ReadingRepository

    @Nested
    @DisplayName("GET /api/v2/bookshelf/popular")
    inner class GetPopular {
        @Test
        @DisplayName("요청하면 → 인기 도서를 조회한다")
        fun getPopularSuccess() {
            SocialReadingSteps.getPopular(mockMvc)
                .andExpect(status().isOk)
        }

        @Test
        @DisplayName("최근 일주일 내 추가 도서가 없으면 → 빈 목록을 반환한다")
        fun getPopularEmpty() {
            SocialReadingSteps.getPopular(mockMvc)
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.result.books").isEmpty)
        }

        @Test
        @DisplayName("인기 도서는 → 최대 10개로 제한된다")
        fun getPopularLimit() {
            for (i in 0 until 12) {
                createReading(true, getMockMember(), "ISBN$i")
            }

            SocialReadingSteps.getPopular(mockMvc)
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.result.books.length()").value(Matchers.lessThanOrEqualTo(10)))
        }
    }

    @Nested
    @DisplayName("GET /api/v2/bookshelf/{readingId}")
    inner class GetReadingDetail {
        @Test
        @DisplayName("공개 독서 상세를 조회하면 → 정보를 반환한다")
        fun getReadingDetailSuccess() {
            val reading = createReading(true, getMockMember(), "ISBN-DETAIL")

            SocialReadingSteps.getDetail(mockMvc, reading.id)
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.result.memberBookId").value(reading.id))
        }

        @Test
        @DisplayName("비공개 + 소유자 조회면 → 200을 반환한다")
        fun getPrivateOwner() {
            val reading = createReading(false, getMockMember(), "ISBN-PRIVATE")

            SocialReadingSteps.getDetail(mockMvc, reading.id)
                .andExpect(status().isOk)
        }

        @Test
        @DisplayName("비공개 + 타인 접근이면 → 404를 반환한다")
        fun getPrivateOther() {
            val other = MemberFixture.aMember()
                .email("o")
                .nickname("o")
                .profileImageUrl("p")
                .providerId("p")
                .providerType(ProviderType.KAKAO)
                .build()
            memberRepository.save(other)
            val reading = createReading(false, other, "ISBN-OTHER")

            SocialReadingSteps.getDetail(mockMvc, reading.id)
                .andExpect(status().isNotFound)
        }

        @Test
        @DisplayName("존재하지 않는 ID면 → 400을 반환한다")
        fun getNotFound() {
            SocialReadingSteps.getDetail(mockMvc, 999L)
                .andExpect(status().isBadRequest)
        }
    }

    private fun createReading(isPublic: Boolean, member: org.veri.be.domain.member.entity.Member, isbn: String): org.veri.be.domain.book.entity.Reading {
        var book = BookFixture.aBook().title("T").image("I").isbn(isbn).build()
        book = bookRepository.save(book)
        val reading = ReadingFixture.aReading()
            .member(member)
            .book(book)
            .isPublic(isPublic)
            .build()
        return readingRepository.save(reading)
    }
}
